lib/vrl/compiler/src/expression/function_closure.rs

```diff
@@ -0,0 +1,160 @@
+use core::ExpressionError;
+
+use crate::expression::Block;
+use crate::parser::Ident;
+use crate::state::Runtime;
+use crate::value::VrlValueConvert;
+use crate::{Context, Expression, Value};
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct FunctionClosure {
+    variables: Vec<Ident>,
+    block: Block,
+}
+
+impl FunctionClosure {
+    pub fn new<T: Into<Ident>>(variables: Vec<T>, block: Block) -> Self {
+        Self {
+            variables: variables.into_iter().map(Into::into).collect(),
+            block,
+        }
+    }
+
+    /// Run the closure to completion, given the provided key/value pair, and
+    /// the runtime context.
+    ///
+    /// The provided values are *NOT* mutated during the run, and no return
+    /// value is provided by this method. See `map_key` or `map_value` for
+    /// mutating alternatives.
+    pub fn run_key_value(
+        &self,
+        ctx: &mut Context,
+        key: &str,
+        value: &Value,
+    ) -> Result<(), ExpressionError> {
+        // TODO: we need to allow `LocalEnv` to take a muable reference to
+        // values, instead of owning them.
+        let cloned_key = key.to_owned();
+        let cloned_value = value.clone();
+
+        let key_ident = self.ident(0);
+        let value_ident = self.ident(1);
+
+        let old_key = insert(ctx.state_mut(), key_ident, cloned_key.into());
+        let old_value = insert(ctx.state_mut(), value_ident, cloned_value);
+
+        self.resolve(ctx)?;
+
+        cleanup(ctx.state_mut(), key_ident, old_key);
+        cleanup(ctx.state_mut(), value_ident, old_value);
```

Comment:
Thanks for the review, I'll add some tests (and potential solutions) on Monday, but here's a quick primer:

> I think this will work fine for key since we are swapping the variable via old_key here. But, I'm not seeing anything here that would clean up zoog.

As the RFC notes, we don't need to clean up `zoog` at runtime, since variable scope is checked at _compile-time_. Meaning, this will return an "undefined variable zoog" error at boot-time, even as _technically_ at runtime the variable actually does exist outside the scope. It doesn't matter because you can never run an invalid program. Compile-time cleanup is guaranteed by the fact that a closure consists of a block, which has its own internal state.

> It's not completely unambiguous from the RFC whether this is the correct behaviour or not.

I agree. Your test covers what I expect to happen.

Reply:
Your comment did make me review the original scoping PR though, as I had a feeling, there was something off with type information of mutated data. I filed a tracking issue to fix this at https://github.com/vectordotdev/vector/issues/12381.

Thanks for the thorough review @StephenWakely 🎉

Reply:
Note, I've resolved the mentioned issues in #12384.

That PR is still in draft, since it requires #12383 to land first. Also, I had to resolve this in a separate PR, as I needed to build the fix on top of #12382, to get access to `for_each`, and test this.

I also updated the compiler improvements tracking issue at #8221 to indicate that we should start investing time in adding unit tests to the compiler, as all of these tests currently live in the test harness, which is unfortunate in cases like these, as it requires a working function to test this case.